lib/ansible/module_utils/common/parameters.py

```diff
@@ -799,53 +722,166 @@ def validate_sub_spec(argument_spec, parameters, prefix='', options_context=None
+                    options_aliases = _handle_aliases(sub_spec, sub_parameters, alias_warnings, alias_deprecations)
                 except (TypeError, ValueError) as e:
                     options_aliases = {}
-                    legal_inputs = None
-                    errors.append(to_native(e))
+                    errors.append(AliasError(to_native(e)))
 
                 for option, alias in alias_warnings:
                     warn('Both option %s and its alias %s are set.' % (option, alias))
 
-                no_log_values.update(list_no_log_values(sub_spec, sub_parameters))
+                try:
+                    no_log_values.update(_list_no_log_values(sub_spec, sub_parameters))
+                except TypeError as te:
+                    errors.append(NoLogError(to_native(te)))
 
-                if legal_inputs is None:
-                    legal_inputs = list(options_aliases.keys()) + list(sub_spec.keys())
-                unsupported_parameters.update(get_unsupported_parameters(sub_spec, sub_parameters, legal_inputs))
+                legal_inputs = _get_legal_inputs(sub_spec, sub_parameters, options_aliases)
+                unsupported_parameters.update(_get_unsupported_parameters(sub_spec, sub_parameters, legal_inputs, options_context))
 
                 try:
-                    check_mutually_exclusive(value.get('mutually_exclusive'), sub_parameters)
+                    check_mutually_exclusive(value.get('mutually_exclusive'), sub_parameters, options_context)
                 except TypeError as e:
-                    errors.append(to_native(e))
+                    errors.append(MutuallyExclusiveError(to_native(e)))
 
-                no_log_values.update(set_defaults(sub_spec, sub_parameters, False))
+                no_log_values.update(_set_defaults(sub_spec, sub_parameters, False))
 
                 try:
-                    check_required_arguments(sub_spec, sub_parameters)
+                    check_required_arguments(sub_spec, sub_parameters, options_context)
                 except TypeError as e:
-                    errors.append(to_native(e))
+                    errors.append(RequiredError(to_native(e)))
 
-                validate_argument_types(sub_spec, sub_parameters, new_prefix, options_context, errors=errors)
-                validate_argument_values(sub_spec, sub_parameters, options_context, errors=errors)
+                _validate_argument_types(sub_spec, sub_parameters, new_prefix, options_context, errors=errors)
+                _validate_argument_values(sub_spec, sub_parameters, options_context, errors=errors)
 
-                checks = [
-                    (check_required_together, 'required_together'),
-                    (check_required_one_of, 'required_one_of'),
-                    (check_required_if, 'required_if'),
-                    (check_required_by, 'required_by'),
-                ]
+                checks = (
```

Comment:
this seems like it should be shared with the main one rather than duplicated here

Reply:
This has tripped me up a few times already: I'll update the code in the `validate()` method but forget to do it in this function. Yet another argument for combining the sub-spec validation. I'll make a variable they can share for now.